content/client-lib-development-guide/features.textile

```diff
@@ -267,7 +268,22 @@ h3(#rest-channel). Channel
 *** @(RSL6a2)@ A set of tests must exist to ensure that the client library provides interoperability for the @extras@ field which is a JSON-encodable object (ie a value that represents a JSON @object@ value and supports serialization to and from JSON text). The test, at a minimum, should publish a message with an @extras@ object such as @{"push":[{"title":"Testing"}]}@ and ensure it is received with an equivalent JSON-encodable object
-** @(RSL6b)@ If, for example, incompatible encryption details are provided or invalid Base64 is detected in the message payload, an error message will be sent to the logger, but the message will still be delivered with last successful decoding and the @encoding@ field. For example, if a message had a decoding of "utf-8/cipher+aes-128-cbc/base64", and the payload was successfully Base64 decoded but the payload could not be encrypted because the @CipherParam@ details were not configured, the message would be delivered with a binary payload and an @encoding@ with the value "utf-8/cipher+aes-128-cbc"
+** @(RSL6b)@ If, for example, incompatible encryption details are provided or invalid Base64 is detected in the message payload, an error message will be sent to the logger, but the message will still be delivered with last successful decoding and the @encoding@ field. For example, if a message had a decoding of "utf-8/cipher+aes-128-cbc/base64", and the payload was successfully Base64 decoded but the payload could not be decrypted because the @CipherParam@ details were not configured, the message would be delivered with a binary payload and an @encoding@ with the value "utf-8/cipher+aes-128-cbc". Additional steps need to be taken if decoding failed on "vcdiff" encoding, see @(RSL6c)@ 
+** @(RSL6c)@ Given "vcdiff" encoded deltas are applied to the previous message published on a channel, when a "vcdiff" encoding fails to be decoded, it makes it impossible for a client to apply subsequent deltas received from that point of failure forward. As such, the client must automatically execute the following recovery procedure in lieu of "RTL7e":#"RTL7e":
+*** @(RSL6c1)@ Log error with code 40018
+*** @(RSL6c2)@ Discard the message
+*** @(RSL6c3)@ Send an @ATTACH@ @ProtocolMessage@ with the channel serial set to the previous message to the message for which "vcdiff" decoding failed to the server, transitioning the channel state to @ATTACHING@, and waiting for a confirmation @ATTACHED@, as per @RTL4c@ and @RTL4f@
+** @(RSL6d)@ The data payload of the last message on each channel, i.e. @Message.data@ has to be stored at all times since it might be needed to decode the "vcdiff" encoding. The last message bookkeeping rules are:
+*** @(RSL6d1)@ If the message encoding string ends in @base64@, the message should be base64-decoded (and the encoding string modified accordingly per "RSL6":#RSL6) before storing the last payload.
```

Comment:
Is the removal of `base64`, on the assumption that it is an encoding step required by the transport, required unconditionally, or only in the case of a text transport?

Reply:
This has been discussed a couple of times and @SimonWoolf provided a good summary of these discussions in Slack:

"I've given my opinion on this before, but will reiterate here for posterity: I would be against this because it adds complexity and is unnecessary.
In particular, in ably-js, the 'obvious' way to do it (looking at whether the currently active transport is text or binary) is subtly buggy, because during the upgrade period there is a period where the active transport is still the comet transport but an incoming message can validly arrive on the upgrade (websocket) transport.
Of course if this were something that was genuinely necessary, there are ways around that: e.g. annotating the message with the format while at the transport level, or decoding base64 if-text-transport as part of deserializing the ProtocolMessage, before it's delivered to the Channel. But it isn't necessary. There is no possibility of a message coming down a binary transport with an encoding string ending in /base64, because even if someone did do some base64-encoding themself and publish a message with the encoding field already set to base64, then (as noted in the call) realtime will have assumed that any base64 encoding in an inbound message was added by the transport and removed it on entering the system. So this would be extra complexity that serves no purpose. (edited) 
simon 5:52 PM
(also I suspect that no customer has or will ever publish a message with an encoding string pre-set. It's one of those things that the system technically supports but no-one ever does because if they want to have app-level encoding info there's no reason for them to use our encoding system rather than have their own (as part of the data object) which they know the client libs aren't going to mess with)"